pkg/util/modules/modules.go

```diff
@@ -0,0 +1,157 @@
+package modules
+
+import (
+	"fmt"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+	"github.com/pkg/errors"
+)
+
+type service func() (services.Service, error)
+
+// module is the basic building block of the application
+type module struct {
+	// dependencies of this module
+	deps []string
+
+	// initFn for this module (can return nil)
+	initFn service
+}
+
+// Manager is a component that initialises modules of the application
+// in the right order of dependencies.
+type Manager struct {
+	modules map[string]module
+}
+
+// NewManager creates a new Manager
+func NewManager() *Manager {
+	return &Manager{
+		modules: make(map[string]module),
+	}
+}
+
+// RegisterModule registers a new module with ModuleManager
+func (m *Manager) RegisterModule(name string, initFn service) {
+	m.modules[name] = module{
+		initFn: initFn,
+	}
+	return
+}
+
+// AddDependency adds a dependency from name(source) to dependsOn(targets)
+func (m *Manager) AddDependency(name string, dependsOn ...string) error {
+	if mod, ok := m.modules[name]; ok {
+		for _, dep := range dependsOn {
+			if _, ok := m.modules[dep]; ok {
+				mod.deps = append(mod.deps, dep)
+			} else {
+				return fmt.Errorf("no such module: %s", dep)
+			}
+		}
```

Comment:
I expected that dependencies will be only checked during `InitModuleServices`, not when adding them. Otherwise modules need to be added in the correct order.

Reply:
I agree we shouldn't bother checking for this here.